mlflow/gateway/providers/togetherai.py

```diff
@@ -0,0 +1,470 @@
+import json
+from typing import Any, AsyncGenerator, AsyncIterable, Dict
+
+from fastapi import HTTPException
+
+from mlflow.exceptions import MlflowException
+from mlflow.gateway.config import RouteConfig, TogetherAIConfig
+from mlflow.gateway.providers.base import BaseProvider, ProviderAdapter
+from mlflow.gateway.providers.utils import rename_payload_keys, send_request, send_stream_request
+from mlflow.gateway.schemas import chat, completions, embeddings
+from mlflow.gateway.utils import strip_sse_prefix
+
+
+class TogetherAIAdapter(ProviderAdapter):
+
+
+    def _scale_repetition_penalty(
```

Comment:
# Description of parameters

## OpenAI (https://platform.openai.com/docs/api-reference/chat/create)
```
frequency_penalty
number or null

Optional
Defaults to 0
Number between -2.0 and 2.0. 
Positive values penalize new tokens based 
on their existing frequency in the text so far, 
decreasing the model's likelihood to repeat the same line verbatim.
```

```
presence_penalty
number or null

Optional
Defaults to 0
Number between -2.0 and 2.0. 
Positive values penalize new tokens based on whether 
they appear in the text so far, increasing the model's likelihood to talk about new topics.
```
## TogetherAI (https://docs.together.ai/reference/chat-completions)

```
repetition_penalty
float
A number that controls the diversity of 
generated text by reducing the likelihood 
of repeated sequences. Higher values decrease repetition.

range ()
```

# My idea  

The parameter ``repetition_penalty`` for ``TogetherAI`` and ``frequency_penalty`` for OpenAI seem to do the same thing, just with different  ranges. If a user uses the ``frequency_penalty`` for a ``TogetherAI`` model, which has a range of (-2,2), they lose a portion of the value range. 

On the other hand the way its structured right now the user can input a value range for ``frequency_penalty`` that matches the ``repetition_penalty`` value range with no problem. If they decide to adhere to the (-2,2)  value range, I think it would be more appropriate to scale it to (-3.402823669209385e+38, 3.402823669209385e+38).

Yet again you are right with outlining that: 
> most likely each model has different sensitivity.

So if we decide to keep this it would be interesting whether I/we can find some info on this. 

Additionally there is the problem of whether matching ``repetition_penalty`` for ``TogetherAI`` to ``frequency_penalty`` instead of ``presence_penalty`` for ``OpenAI``. After reading the docs a bit my choice seems to be the most appropriate. 

I wasn't 100% sure of this when I did it. Would love to hear your thoughts!

Reply:
Thank you for the detailed explanation!

> If they decide to adhere to the (-2,2) value range, I think it would be more appropriate to scale it to (-3.402823669209385e+38, 3.402823669209385e+38).

My concern is that we don't know the exact mapping function between two ranges. If the assumption that OpenAI shrinks the infinite value range (-3.402823669209385e+38, 3.402823669209385e+38) into (-2, 2), it must not be a linear projection, rather some sort of squashing function like sigmoid, otherwise it would be too sensitive to the lower side of the range. For example, if we do linear projection, the tiny increase of the `frequency_penalty`like `0.0000001`, would shift the value passed to the underyling model 3e+30. I'm not super familiar with TogetherAI but I doublt it is common practice to pass such large number to the parameter.

Also, it seems the default value of `repetition_penalty` is `1` in TogetherAI. That would be extremely small number if projected to (-2, 2) and most likely have underflow problem.

> Additionally there is the problem of whether matching repetition_penalty for TogetherAI to frequency_penalty instead of presence_penalty for OpenAI.

Indeed, I think it's totally fine to have the `repetition_penalty` as it is. Aside from the [standard parameters](https://mlflow.org/docs/latest/llms/deployments/index.html#completions), we don't put restriction for what query parameters to be passed to the underlying model provider. If there would be a confusion by the renaming, I'd just expose the original parameter unchanged.

Reply:
> My concern is that we don't know the exact mapping function between two ranges. 

Yeap you are totally right on this and the above. I will remove the function. 

> Aside from the [standard parameters](https://mlflow.org/docs/latest/llms/deployments/index.html#completions), we don't put restriction for what query parameters to be passed to the underlying model provider.

That is very useful to know! Thanks!